From cb5a3711c6bc5928865916175bfb388829348822 Mon Sep 17 00:00:00 2001
From: Michael Weghorn <m.weghorn@posteo.de>
Date: Mon, 22 Sep 2025 14:46:12 +0200
Subject: QQuickTextInput a11y: Allow modifying text selection via a11y API

Similar to the previous commit that introduced
QAccessibleQuickTextEdit, introduce QAccessibleQuickTextInput
and implement the logic to modify selection.

This allows changing the selection for QQuickTextInput
and its QQuickTextField subclass via the platform
accessibility API, e.g. the AT-SPI Text interface's
SetSelection method.

Extend the existing TextInput autotest accordingly.

Task-number: QTBUG-139943
Task-number: QTBUG-140441
Change-Id: Ie05c6b0cf3431aabb4b337dbfa07c40c56660e2b
---
 src/quick/CMakeLists.txt                      |  1 +
 .../accessible/qaccessiblequicktextinput.cpp  | 31 ++++++++++++++
 .../accessible/qaccessiblequicktextinput_p.h  | 42 +++++++++++++++++++
 .../accessible/qquickaccessiblefactory.cpp    |  4 ++
 .../qquickaccessible/tst_qquickaccessible.cpp |  9 ++++
 5 files changed, 87 insertions(+)
 create mode 100644 src/quick/accessible/qaccessiblequicktextinput.cpp
 create mode 100644 src/quick/accessible/qaccessiblequicktextinput_p.h

diff --git x/qtdeclarative/src/quick/CMakeLists.txt y/qtdeclarative/src/quick/CMakeLists.txt
index 087087982c24ec1eeb08353aabbfe889175694e5..8103f939274a728234441b8d340e257f419f020b 100644
--- x/qtdeclarative/src/quick/CMakeLists.txt
+++ y/qtdeclarative/src/quick/CMakeLists.txt
@@ -570,6 +570,7 @@ qt_internal_extend_target(Quick CONDITION QT_FEATURE_accessibility
     SOURCES
         accessible/qaccessiblequickitem.cpp accessible/qaccessiblequickitem_p.h
         accessible/qaccessiblequicktextedit.cpp accessible/qaccessiblequicktextedit_p.h
+        accessible/qaccessiblequicktextinput.cpp accessible/qaccessiblequicktextinput_p.h
         accessible/qaccessiblequickview.cpp accessible/qaccessiblequickview_p.h
         accessible/qquickaccessiblefactory.cpp accessible/qquickaccessiblefactory_p.h
     LIBRARIES
diff --git x/qtdeclarative/src/quick/accessible/qaccessiblequicktextinput.cpp y/qtdeclarative/src/quick/accessible/qaccessiblequicktextinput.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..23132e0cedc2e11d8f658bc6b61de0a71332d5b7
--- /dev/null
+++ y/qtdeclarative/src/quick/accessible/qaccessiblequicktextinput.cpp
@@ -0,0 +1,31 @@
+// Copyright (C) 2025 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#include "qaccessiblequicktextinput_p.h"
+
+QT_BEGIN_NAMESPACE
+
+#if QT_CONFIG(accessibility)
+
+QAccessibleQuickTextInput::QAccessibleQuickTextInput(QQuickTextInput *textEdit)
+    : QAccessibleQuickItem(textEdit)
+{
+}
+
+void QAccessibleQuickTextInput::removeSelection(int selectionIndex)
+{
+    if (selectionCount() == 1 && selectionIndex == 0) {
+        const int cursorPos = textInput()->cursorPosition();
+        textInput()->select(cursorPos, cursorPos);
+    }
+}
+
+void QAccessibleQuickTextInput::setSelection(int selectionIndex, int startOffset, int endOffset)
+{
+    if (selectionIndex == 0)
+        textInput()->select(startOffset, endOffset);
+}
+
+#endif // accessibility
+
+QT_END_NAMESPACE
diff --git x/qtdeclarative/src/quick/accessible/qaccessiblequicktextinput_p.h y/qtdeclarative/src/quick/accessible/qaccessiblequicktextinput_p.h
new file mode 100644
index 0000000000000000000000000000000000000000..1ba46f0f601d087b839e320b480f2fd5379b9ec8
--- /dev/null
+++ y/qtdeclarative/src/quick/accessible/qaccessiblequicktextinput_p.h
@@ -0,0 +1,42 @@
+// Copyright (C) 2025 The Qt Company Ltd.
+// SPDX-License-Identifier: LicenseRef-Qt-Commercial OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#ifndef QACCESSIBLEQUICKTEXTINPUT_H
+#define QACCESSIBLEQUICKTEXTINPUT_H
+
+//
+//  W A R N I N G
+//  -------------
+//
+// This file is not part of the Qt API.  It exists purely as an
+// implementation detail.  This header file may change from version to
+// version without notice, or even be removed.
+//
+// We mean it.
+//
+
+#include "qaccessiblequickitem_p.h"
+
+#include <QtQuick/private/qquicktextinput_p.h>
+
+QT_BEGIN_NAMESPACE
+
+#if QT_CONFIG(accessibility)
+
+class Q_QUICK_EXPORT QAccessibleQuickTextInput : public QAccessibleQuickItem
+{
+public:
+    QAccessibleQuickTextInput(QQuickTextInput *textEdit);
+
+    void removeSelection(int selectionIndex) override;
+    void setSelection(int selectionIndex, int startOffset, int endOffset) override;
+
+private:
+    QQuickTextInput *textInput() const { return static_cast<QQuickTextInput *>(item()); }
+};
+
+#endif // accessibility
+
+QT_END_NAMESPACE
+
+#endif // QACCESSIBLEQUICKTEXTINPUT_H
diff --git x/qtdeclarative/src/quick/accessible/qquickaccessiblefactory.cpp y/qtdeclarative/src/quick/accessible/qquickaccessiblefactory.cpp
index f8bce6a663040b850aef08a52202ae07b7ba3f7f..88614307557c73399bd3ae885f211e1a70fa0cf1 100644
--- x/qtdeclarative/src/quick/accessible/qquickaccessiblefactory.cpp
+++ y/qtdeclarative/src/quick/accessible/qquickaccessiblefactory.cpp
@@ -6,8 +6,10 @@
 #include "qaccessiblequickview_p.h"
 #include "qaccessiblequickitem_p.h"
 #include "qaccessiblequicktextedit_p.h"
+#include "qaccessiblequicktextinput_p.h"
 #include <QtQuick/private/qquickitem_p.h>
 #include <QtQuick/private/qquicktextedit_p.h>
+#include <QtQuick/private/qquicktextinput_p.h>
 
 QT_BEGIN_NAMESPACE
 #if QT_CONFIG(accessibility)
@@ -18,6 +20,8 @@ QAccessibleInterface *qQuickAccessibleFactory(const QString &classname, QObject
         return new QAccessibleQuickWindow(qobject_cast<QQuickWindow *>(object));
     if (classname == QLatin1String("QQuickTextEdit"))
         return new QAccessibleQuickTextEdit(qobject_cast<QQuickTextEdit *>(object));
+    if (classname == QLatin1String("QQuickTextInput"))
+        return new QAccessibleQuickTextInput(qobject_cast<QQuickTextInput *>(object));
     if (classname == QLatin1String("QQuickItem")) {
         QQuickItem *item = qobject_cast<QQuickItem *>(object);
         Q_ASSERT(item);
diff --git x/qtdeclarative/tests/auto/quick/qquickaccessible/tst_qquickaccessible.cpp y/qtdeclarative/tests/auto/quick/qquickaccessible/tst_qquickaccessible.cpp
index 1f7e27e0404311d70120b03ea8eddbfee24beb15..cc9aece71b9d11d3145e90473f45091a81430a21 100644
--- x/qtdeclarative/tests/auto/quick/qquickaccessible/tst_qquickaccessible.cpp
+++ y/qtdeclarative/tests/auto/quick/qquickaccessible/tst_qquickaccessible.cpp
@@ -512,6 +512,15 @@ void tst_QQuickAccessible::basicPropertiesTest()
     auto newText = QString("a new text");
     textInput->setText(QAccessible::Value, newText);
     QCOMPARE(textInput->text(QAccessible::Value), newText);
+    QCOMPARE(textInterface->selectionCount(), 0);
+    textInterface->setSelection(0, 1, 4);
+    QCOMPARE(textInterface->selectionCount(), 1);
+    int selectionStartOffset = 0, selectionEndOffset = 0;
+    textInterface->selection(0, &selectionStartOffset, &selectionEndOffset);
+    QCOMPARE(selectionStartOffset, 1);
+    QCOMPARE(selectionEndOffset, 4);
+    textInterface->removeSelection(0);
+    QCOMPARE(textInterface->selectionCount(), 0);
 
     // TextEdit
     QAccessibleInterface *textEdit = item->child(3);
